### PR TITLE
faster code

### PR DESCRIPTION
**Pull request by guliashvili**
Sometimes realloc will save time by not allocating new space and just extending current.

My stockfish			Original Stockfish
4752			4794
4802			4828
4302			4830
4861			4930
4793			4847
4836			4828

average	4724.333333		average = 	4842.833333
stdevP	191.9311219		stdevp = 	42.04131566

High stdev of my stockfish is predictable. realloc will work only in lucky cases(when array is extendable so no free will be called) and in unlucky cases it would be same.

### CODE REVIEW

File: src/tt.cpp [hunk 1]
```diff
@@ -39,14 +39,15 @@ void TranspositionTable::resize(size_t mbSize) {
...
-  free(mem);
-  mem = calloc(clusterCount * sizeof(Cluster) + CacheLineSize - 1, 1);
+  mem = realloc(mem, clusterCount * sizeof(Cluster) + CacheLineSize - 1);
 
   if (!mem)
   {
       std::cerr << "Failed to allocate " << mbSize
                 << "MB for transposition table." << std::endl;
       exit(EXIT_FAILURE);
+  }else{
```
Comment (mstembera) [inline src/tt.cpp]: You could just call clear() at the end after table is assigned.

Comment (guliashvili) [inline src/tt.cpp]: Thanks

File: src/tt.cpp [hunk 2]
```diff
@@ -48,6 +47,8 @@ void TranspositionTable::resize(size_t mbSize) {
                 << "MB for transposition table." << std::endl;
       exit(EXIT_FAILURE);
   }
+  memset(mem, 0, clusterCount * sizeof(Cluster) + CacheLineSize - 1);
```
Comment (mstembera) [inline src/tt.cpp]: Sorry, I meant clear() which is already a member of TranspositionTable so u don't have to add another manual memset() call.  Just call clear() at the very bottom of resize() after the table=... line.

Comment (guliashvili) [inline src/tt.cpp]: clear cleans table not mem

Comment (guliashvili) [inline src/tt.cpp]: I'm not sure about that
at least until now calloc would clean whole mem 
clear cleans table not mem. seems like table is kind of shifted one

Comment (mstembera) [inline src/tt.cpp]: Yes the difference between mem and table is only cache line alignment.  We don't care what's in the first few bytes of mem that will never be accessed.

Comment (guliashvili) [inline src/tt.cpp]: done. thanks

**Comment by lucasart** [comment]
Optimizing this is useless. Hash is allocated once, then you don't care. It's fast enough as it it.

But the real reason why `realloc` is superior to `calloc`, is that you preserve the existing hash. People using SF in long analysis may find this useful. For *that* reason the patch is a good idea IMO.

**Comment by mcostalba** [comment]
@lucasart in this case you have to remove the clear() call, but this makes
things a bit slippery IMO

**Comment by guliashvili** [comment]
@lucasart 
If its called only once then why was there function free?

**Comment by mcostalba** [comment]
What Lucas means is that it is called before the game starts  so the speed
is not an issue.

**Comment by guliashvili** [comment]
Aha got it.
Mm in any case I think realloc should be used instead of free calloc

**Comment by mcostalba** [comment]
In the past I experimented with a std ::vector for the TT table,
unfortunately it may seems to
be a small speed regression, I think mainly due to the cache alignment....
Maybe it's about time to respin that idea...

**Comment by guliashvili** [comment]
I think STL  will slow things down. Of course it's less error prone and blah blah but this project needs speed :-D

**Comment by zamar** [comment]
As others have pointed out, the code is not in the hot path. It doesn't have any impact to speed.

free() + malloc() is just fine, I can't see a valid reason for the change. realloc() has it's own complexities.

And this project is not only about speed. It's about speed and clean code. The latter is very important in the long run.

**Comment by mstembera** [comment]
Does anyone know if using realloc() reduces memory fragmentation compared to calling free() and calloc()?

**Comment by lucasart** [comment]
@mcostalba: ah yes, calling `clear()` after `realloc()` is silly because it negates the only benefit of `realloc()`. so this patch is only about a speedup, which is totally misguided. So I guess the pull request can be closed.

**Comment by lucasart** [comment]
To clarify what I mean with some actual code, see `TT::alloc()` in DiscoCheck:
 https://github.com/lucasart/chess/blob/master/src/tt.cc

**Comment by guliashvili** [comment]
Yeap you are right guys. thanks. closed.

**Pull request closed without merging**